packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol

```diff
@@ -374,50 +263,238 @@ library ExchangeBlocks
+        uint offset = 4 + 32 + 32 + 32;
+        // The length of the auxiliary data needs to match the number of conditional transfers
+        uint numConditionalTransfers = data.bytesToUint32(offset);
+        require(auxiliaryData.length == numConditionalTransfers * 4, "INVALID_AUXILIARYDATA_LENGTH");
+        offset += 4;
+
+        uint24 operatorAccountID = data.bytesToUint24(offset);
+        offset += 3;
+
+        if (numConditionalTransfers > 0) {
+            require(S.onchainDataAvailability, "CONDITIONAL_TRANSFERS_REQUIRE_OCDA");
+        }
+
+        // Run over all conditional transfers
+        uint previousTransferOffset = 0;
+        for (uint i = 0; i < auxiliaryData.length; i += 4) {
+            // auxiliaryData contains the transfer index (4 bytes) for each conditional transfer
+            // in ascending order.
+            uint transferOffset = offset + auxiliaryData.bytesToUint32(i) * 14 + 14;
+            require(transferOffset > previousTransferOffset, "INVALID_AUXILIARYDATA_DATA");
+
+            // Get the transfer data
+            uint transferData;
+            assembly {
+                transferData := and(mload(add(data, transferOffset)), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFF)
+            }
+
+            // Check that this is a conditional transfer
+            uint transferType = (transferData >> 104) & 0xFF;
+            require(transferType == 1, "INVALID_AUXILIARYDATA_DATA");
+
+            // Update the onchain state
+            consumeConditionalTransfer(
+                S,
+                operatorAccountID,
+                transferData
+            );
+
+            previousTransferOffset = transferOffset;
+        }
+    }
+
+    function consumeConditionalTransfer(
+        ExchangeData.State storage S,
+        uint24 operatorAccountID,
+        uint transferData
+        )
+        private
+    {
+        // Extract the transfer data
+        uint24 fromAccountID = uint24((transferData >> 80) & 0xFFFFFF);
+        uint24 toAccountID = uint24((transferData >> 56) & 0xFFFFFF);
+        uint16 tokenID = uint16((transferData >> 48) & 0xFF);
+        uint amount = ((transferData >> 24) & 0xFFFFFF).decodeFloat(24);
+        uint16 feeTokenID = uint16((transferData >> 16) & 0xFF);
+        uint feeAmount = ((transferData >> 0) & 0xFFFF).decodeFloat(16);
+
+        // Update onchain approvals
+        if (amount > 0) {
+            S.approvedTransferAmounts[fromAccountID][toAccountID][tokenID] =
+                S.approvedTransferAmounts[fromAccountID][toAccountID][tokenID].sub(amount);
+            emit ConditionalTransferConsumed(fromAccountID, toAccountID, tokenID, amount);
+        }
+        if (feeAmount > 0) {
+            S.approvedTransferAmounts[fromAccountID][operatorAccountID][feeTokenID] =
+                S.approvedTransferAmounts[fromAccountID][operatorAccountID][feeTokenID].sub(feeAmount);
+            emit ConditionalTransferConsumed(fromAccountID, operatorAccountID, feeTokenID, feeAmount);
```

Comment:
So consume one single transfer may end up with one additional fee transfer. Does it mean for fastwithdrawl, an extra approved is required for paying the operator?

Is it possible that we whitelist the operator so paying the operator doesn't require approval at all?

Reply:
If there are indeed 2 different transfers necessary then 2 separate approvals would be needed. Though I think it's unlikely that a different offchain transfer for the fee payment will be used in practice because it'll probably make a lot of sense to just pay the fee in the same token if other funds are sent offchain. 

For the fast withdrawal case, it makes a lot of sense to just keep a small part of the user's withdrawal amount (comparable to how centralized exchanges do). This way the no separate fee payment is necessary and a single approval is sufficient.

I'm pretty sure it's not possible to do any whitelisting because I don't see how we could limit the amount he could transfer out of users' account, even if it is a "fee payment". It's also necessary that some state is changed onchain to mark the transfer as done, otherwise it could be reused as many times the operator wants.

Reply:
I strongly prefer the fees are paid in the exact same withdrawal token!

Reply:
Sure, the option is just there to pay the fee in another token. It's up to the exchange to decide what it wants to do.